Lib/ctypes/util.py

```diff
@@ -80,6 +141,28 @@ def find_library(name):
                 continue
         return None
 
+    import ctypes
+
+    # https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dyld.3.html
+    _libc = ctypes.CDLL(find_library("c"))
+    _dyld_get_image_name = _libc["_dyld_get_image_name"]
+    _dyld_get_image_name.restype = ctypes.c_char_p
+
+    def dllist():
+        try:
+            libraries = []
+            num_images = _libc._dyld_image_count()
+
+            # start at 1 to skip executable
+            for i in range(1, num_images):
+                raw_name = _dyld_get_image_name(i)
+                name = os.fsdecode(raw_name)
+                libraries.append(name)
+
+            return libraries
+        except Exception:
+            return None
```

Comment:
I still think this is bad design, we should let the error propagate. As you said, there's not much a user could do if an exception occurs, but that goes for a `None` return as well.

FWIW, if we do go the route of returning `None` (although, that would take convincing), catching `Exception` is the wrong way to do it here. If something unrelated goes wrong (e.g. out of memory, some internal corruption, or a million other things that could go wrong), *that* should definitely propagate -- move the `except Exception` to specific lines, and expect specific exceptions.

Reply:
To me the difference arises if I'm in an application where I want to treat errors in the same way as "couldn't find the library I'm looking for".

If we return None, then I can write 

`'libfoo.so' in (dllist() or [])`

If dllist can throw an exception, this same logic becomes something like 
```python
dlls = []
try:
	dlls = dllist()
except:
	pass

if 'libfoo.so' in dlls:
	...
```
There's no real information gained in the second case, but handling an exception is much more verbose than handling a value for this.

---

Conversely, if we _don't_ want to treat it like this, you can write this in the None case:

```python
dlls = dllist()
if dlls is None:
	raise RuntimeError("could not determine if libfoo was loaded")

if 'libfoo.so' in dlls:
	...
```

Or this, in the other case:

```python
dlls = []
try:
	dlls = dllist()
except Exception as e:
	raise RuntimeError("could not determine if libfoo was loaded") from e

if 'libfoo.so' in dlls:
	...
```

This case is _also_ more verbose, and the only advantage it has is the `from e`, which may or may not contain any meaningful information for the actual user.

---

It's also the same behavior that `ctypes.util.find_library` has, e.g.:
https://github.com/python/cpython/blob/e2f2dc708eae89f41e328501b5ea7c97b8e39907/Lib/ctypes/util.py#L335-L336

Reply:
My main issue with a `None` return is that you *force* the user to suppress any exception that occurred -- some may want to write to logs, reraise it, or in your case, catch it and assign to an empty list. As you've shown, it's pretty trivial to catch the exception and assign the result to something more useful, but I don't think we should force that choice on everyone.

`find_library` is a bit different to me, because when I think of finding something, it's sort of implied that the thing might not have been found -- *that's* what the `None` return is for, not for indicating that an error occurred.

Reply:
Would you propose we also raise unconditionally in the "supported platform" case?

Reply:
I think we could return `None` in some cases, but let errors propagate that are unrelated.

Reply:
That would make callers check both cases, which seems like a worst-of-both-worlds

Reply:
I did some more reading on the various APIs that we call, and the only one that can actually "fail" is Windows. MacOS's API might return NULL for the name of some modules, so I'm handling that explicitly, but otherwise we don't need these `except Exception` blocks 

The Windows function only fails in the case
> If the module list in the target process is corrupted or not yet initialized

So while I have left the error handling code there, I do not believe you would ever encounter it while managing to keep Python itself up and running...